Invite organization email and assign points if signed in.

sidd190 opened the pull request:
# Organization Invite Points System Implementation

Fixes #4550 

## Summary

Successfully implemented the points awarding logic for organization invites. When a user invites an organization and that organization registers via the referral link, the inviting user receives 5 points.

**Current Invite functionality**

https://github.com/user-attachments/assets/891a989c-2280-46e5-8aa2-ac0ac6738fee

**Implemented invite functionality**

[Screencast from 2025-11-01 00-05-42.webm](https://github.com/user-attachments/assets/f1aabb75-a926-4507-a457-39b2daa3f70f)

## Changes Made

### 1. Fixed Organization Registration View (`website/views/company.py`)
- **Added imports**: `InviteOrganization`, `Points`
- **Added points logic**: After successful organization creation, check for referral code in session
- **Award 5 points**: Create Points record for the inviting user
- **Update invite record**: Mark `points_awarded=True` and link to created organization
- **Prevent double points**: Only award points if `points_awarded=False`

### 2. Improved Invite View (`website/views/core.py`)
- **Fixed placeholder issue**: Removed creation of fake placeholder invites
- **Generate sample links**: Use UUID for display-only referral links
- **Better context**: Added proper template variables for email generation
- **Security improvement**: Sample links can't be used to claim points

### 3. Updated Template (`website/templates/invite.html`)
- **Sample link indicator**: Different messages for real vs sample referral links
- **User context**: Show appropriate messages based on login status

### 4. Consolidated URL Routing (`blt/urls.py`)
- **Single invite view**: Use `invite_organization` function from core.py
- **Removed duplicate**: Eliminated unused `InviteCreate` class from user.py

### 5. Comprehensive Tests (`website/test_organization_invite.py`)
- **Points awarding**: Test that 5 points are awarded on successful registration
- **No double points**: Test that same referral code can't be used twice
- **Invalid codes**: Test that invalid referral codes don't award points
- **Sample links**: Test that sample links don't award points
- **Anonymous users**: Test behavior for non-logged-in users

## How It Works

1. **User creates invite**: Logged-in user fills invite form → `InviteOrganization` record created
2. **Referral link generated**: Unique referral code appended to organization registration URL
3. **Organization visits link**: Referral code stored in session
4. **Organization registers**: Registration form submitted successfully
5. **Points awarded**: System finds matching invite record and awards 5 points to sender
6. **Invite updated**: Mark as used and link to created organization

## Key Features

**5 points per successful invite**  
**Prevents duplicate point awards**  
**Secure referral tracking**  
**Works for logged-in users only**  
**Sample links for demonstration**  
**Comprehensive test coverage**  

## Security Considerations

- Only real invite records (not sample links) can award points
- Referral codes are unique UUIDs
- Points only awarded once per invite
- Session-based referral tracking prevents tampering
- Invalid referral codes are handled gracefully

## Testing

Run the test suite to verify functionality:
```bash
python manage.py test website.test_organization_invite
```

The implementation includes tests for all major scenarios including edge cases and security considerations.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Organization invite flow: generate professional, customizable invitation emails with preview, per-field copy, "Copy All" and "Open in Email Client", plus referral links and referral-based points when an invited organization registers.

* **Bug Fixes / UX**
  * Improved form messaging, field-specific error reporting, conditional copy for logged-in vs guest users, robust referral/session handling and safer URL parsing.

* **Tests**
  * End-to-end coverage for invite workflow, referral handling, and point-awarding logic.

* **Documentation**
  * Expanded invite docs covering workflow, email content, accessibility, and chatbot guidance.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Copilot committed:
Initial plan

Copilot committed:
Enhanced invite organization functionality with professional email generation

Co-authored-by: DonnieBLT <128622481+DonnieBLT@users.noreply.github.com>

Copilot committed:
Add comprehensive documentation and testing for enhanced invite feature

Co-authored-by: DonnieBLT <128622481+DonnieBLT@users.noreply.github.com>

sidd190 committed:
Fixed some errors

sidd190 committed:
Added organisation invite, with/without login, awarding 5 points per joinee with referral

sidd190 committed:
Added comprehensive documentation and testing

sidd190 committed:
Merge remote-tracking branch 'upstream/main' into invite-organization-email

sidd190 committed:
Code Corrections, update per review

sidd190 committed:
Fixed data inconsistency causing tests to fail

sidd190 committed:
Removed minor error and handled null-edge cases

sidd190 committed:
remove cookies

sidd190 commented on website/views/core.py:
```diff
@@ -2896,3 +2897,117 @@ def get_context_data(self, **kwargs):
+            base_url = request.build_absolute_uri(reverse("register_organization"))
+            default_referral = request.build_absolute_uri(reverse("register_organization"))
+            referral_link = context.get("referral_link", default_referral)
+            referral_link = f"{base_url}?ref={invite_record.referral_code}"
```
Can i remove  referral_link = f"{base_url}?ref={invite_record.referral_code}"
instead of the default referral logic?

sidd190 committed:
Update referral generation logic

sidd190 committed:
Clear sessions with correct scope, other changes in referral logic

sidd190 committed:
Company.py reformatted, pre-commit tests pass

sidd190 committed:
Added nested transaction for referral processing

DonnieBLT commented on INVITE_ORGANIZATION_FEATURE.md:
remove this

DonnieBLT commented on .gitignore:
```diff
@@ -15,6 +15,7 @@ geckodriver.log
 geckodriver
 chromedriver
 requirements.txt
+cookies.txt
```
remove this

DonnieBLT commented on website/templates/leaderboard_global.html:
```diff
@@ -120,28 +130,32 @@ <h1 class="page-header">Global Leaderboard</h1>
                                                  height="50px">
                                         {% else %}
-                                            <img src="{% gravatar_url leader.reviews__reviewer__user__email 50 %}"
+                                            <img src="{% gravatar_url leader.reviewer__user__email 50 %}"
```
don't mix other issues please

sidd190 replied:
Reverted changes.

sidd190 committed:
Removed extras, reverted leaderboard changes, and updated files as per review

sidd190 committed:
Added robust null value handling and updation as per review

sidd190 commented on test_invite.py:
Yes I was using it initially locally to double check the functionality and added by mistake.

sidd190 commented on website/templates/invite.html:
```diff
@@ -5,66 +5,176 @@
+                        {% endif %}
                         <p>
-                            {% trans "Invite an organization to join and earn" %} <span class="font-bold text-red-500/70">{% trans "+5 points" %}</span> {% trans "when they do." %}
+                            {% trans "Generate a professional invitation email with detailed information about BLT's benefits for organizations." %}
```
Ahh, yeah makes sense. I'll update it.

sidd190 committed:
Removed empty file, updated code as per reviews

sidd190 committed:
Move email body to templates and update tests for transaction.oncommit

sidd190 committed:
Merge branch 'main' into invite-organization-email

sidd190 committed:
Update as per code reviews

sidd190 committed:
Update user.py for better handling

sidd190 committed:
Merge branch 'main' into invite-organization-email

sidd190 committed:
Updated with consolidations and fixed data context flow

sidd190 committed:
Merge remote changes from invite-organization-email

sidd190 committed:
Fix HTML attribute syntax and improve invite validation logic

- Fix malformed required attribute in invite.html input field
- Add validation for partial submissions from unauthenticated users
- Prevent duplicate invites using get_or_create pattern
- Add error handling for template rendering failures

sidd190 committed:
Fix linting for invite.html
